lib/semantics/check-do-concurrent.cc

```diff
@@ -215,32 +216,91 @@ class FindDoConcurrentLoops {
   FindDoConcurrentLoops(parser::Messages &messages) : messages_{messages} {}
   template<typename T> constexpr bool Pre(const T &) { return true; }
   template<typename T> constexpr void Post(const T &) {}
-  template<typename T> constexpr bool Pre(const parser::Statement<T> &) {
-    return false;
-  }
-  bool Pre(const parser::DoConstruct &doConstruct) {
-    if (std::get<std::optional<parser::LoopControl>>(
-            std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t)
-                .statement.t)
-            .has_value() &&
-        std::holds_alternative<parser::LoopControl::Concurrent>(
-            std::get<std::optional<parser::LoopControl>>(
-                std::get<parser::Statement<parser::NonLabelDoStmt>>(
-                    doConstruct.t)
-                    .statement.t)
-                ->u)) {
-      DoConcurrentEnforcement doConcurrentEnforcement{messages_};
-      parser::Walk(
-          std::get<parser::Block>(doConstruct.t), doConcurrentEnforcement);
-      DoConcurrentLabelEnforce doConcurrentLabelEnforce{
-          messages_, doConcurrentEnforcement.labels()};
-      parser::Walk(
-          std::get<parser::Block>(doConstruct.t), doConcurrentLabelEnforce);
+  void Post(const parser::DoConstruct &doConstruct) {
+    auto &doStmt{
+        std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t)};
+    if (std::get<std::optional<parser::LoopControl>>(doStmt.statement.t)
```

Comment:
The expression `std::get<std::optional<parser::LoopControl>>(doStmt.statement.t)` is repeated 4 times in this function. That makes it hard to read.